sycl/plugins/opencl/pi_opencl.cpp

```diff
@@ -1397,6 +1480,8 @@ pi_result piPluginInit(pi_plugin *PluginInit) {
   // PI interface supports higher version or the same version.
   strncpy(PluginInit->PluginVersion, SupportedVersion, 4);
 
+  ExtFuncsCaches = new ExtFuncsCachesT;
```

Comment:
Should we use a smart pointer? https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#r3-a-raw-pointer-a-t-is-non-owning

Reply:
It might be an option if the following drawback is resolved. Smart pointer (both `shared_ptr` and `unique_ptr`) are not trivially destructible. Having a smart pointer here may lead to potential mem leak or anything else related to order of destructors calls.

Probably, @alexbatashev  has smth to comment here if I'm wrong.

Reply:
> Having a smart pointer here may lead to potential mem leak or anything else related to order of destructors calls.

Could you clarify the scenario when "mem leak or anything else related to order of destructors calls" happen and how using a raw pointer solves the problems in such scenarios?

From POV, you can always call `release` method when you call `delete` for a raw pointer and have additional mem leak protection for the cases when `delete` is not called.

Reply:
The mem-leak and destructor ordering I refer here are not related to the owned object, i.e. cache, but to the smart pointer itself. `shared_ptr` has two atomic counters. My blind guess is they're in heap. Both `unique_ptr` and `shared_ptr` have a 'complex' d-tor which has to check if the pointer owns the object.

Sure thing. When smart pointer is used I'll put call to `release` method in `piTearDown`. Bearing this in mind, is there any real need for smart pointer as I won't use its features?

Reply:
If we change the type of `ExtFuncsCaches` to a `unique_ptr`, we may have a problem since `unique_ptr` can be destructed(as a regular global var) earlier than the latest call to an API which uses this var.

Reply:
How is that possible? Isn't calling functions from unloaded sharing library is UB?

Reply:
> How is that possible? Isn't calling functions from unloaded sharing library is UB?

A global object can be destroyed earlier then a library is unloaded.

Reply:
Right, so I expected that plug-in has a single global object managing the lifetime of plug-in data ("plug-in context") and after this global object is destroyed, there can't be any "calls to an API which uses this var".

Reply:
>> there can't be any "calls to an API which uses this var".

How this can be guaranteed?